swim/state_transitions.go

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package swim
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/uber-common/bark"
+	"github.com/uber/ringpop-go/logging"
+	"github.com/uber/ringpop-go/util"
+)
+
+// subject is an interface to define the subject (eg. member) to transition state for
+type subject interface {
+	address() string
+	incarnation() int64
+}
+
+type transitionTimer struct {
+	*time.Timer
+
+	// state represents the state the subject was in when the transition was scheduled
+	state string
+
+	// transition contains the transition to execute
+	transition func()
```

Comment:
It is actually not being used :), I put it on here at first because I named the struct `transitionTimer`. It had a timer already, but no transition. Now that all is implemented I will remove this field again since it is not used.